### PR TITLE
Add Swipe to Carousel Elements 

### LINKED ISSUE #59: Support touch/swipe on mobile for carousels.

### PR DESCRIPTION
**Pull request by AlexMNet**
Add the ability to swipe left/right on carousel elements on mobile sized devices using the Hammer JS Library
Hammer JS Library installed as well as the HammerJS Types for typescript.

### CODE REVIEW

**Comment by AlexMNet** [comment]
It's easy to accidentally scroll up or down the page when trying to swipe through the carousel. Need to look into fixing that.

File: src/carousel.ts [hunk 1]
```diff
@@ -22,6 +24,34 @@ function carousel(id: string, curChild=0) {
         centerCard(curChild);
     });
 
+
+    //Enable Swipe on mobile devices only
+    hammer.set({enable: windowSize <= 768 ? true : false})
```
Comment (mckoss) [inline src/carousel.ts]: Don't we want this to work on an iPad too?

File: src/carousel.ts [hunk 2]
```diff
@@ -1,10 +1,14 @@
+import Hammer from 'hammerjs';
```
Comment (mckoss) [inline src/carousel.ts:1]: This is the correct place to import.  We generally don't want to depend on global variables for imports.

File: src/carousel.ts [hunk 3]
```diff
@@ -22,6 +26,34 @@ function carousel(id: string, curChild=0) {
...
+    hammer.set({enable: windowSize <= 768 ? true : false})
+    //Detect swipe left or right on divParent
+    hammer.on("swipeleft swiperight", function(e: any) {
+        const {type, isFinal} = e;
+
+        //Event listener fires mutiple times. Waits until events are done.
+        if(isFinal){
+            //Check if user swipes right (image moves right)
+            if(type === 'swiperight'){
+                if(curChild === 0){
```
Comment (mckoss) [inline src/carousel.ts]: We are getting a lot of redundant code here.  Would be best to consolidate.  Perhaps replace the addEventListeners on divLeft and divRight with:

() => scrollBy(-1)  and () => scrollBy(1)

```
function scrollBy(n: number) {
  const nextChild = curChild + n;
  if (nextChild < 0 || nextChild > numChildren) {
     return;
  }
  curChild = nextChild;
  centerCard(curChild);
 showControls();
```
  
}

File: src/carousel.ts [hunk 4]
```diff
@@ -22,6 +26,34 @@ function carousel(id: string, curChild=0) {
         centerCard(curChild);
     });
 
+
+    //Enable Swipe on mobile devices only
+    hammer.set({enable: windowSize <= 768 ? true : false})
+    //Detect swipe left or right on divParent
+    hammer.on("swipeleft swiperight", function(e: any) {
```
Comment (mckoss) [inline src/carousel.ts]: Prefer arrow function:

(e) => {

any is not a great type declaration.  Hammer should be typed in such a way as to
not need an explicit declaration here - it should be inferred and have Hammer's
properties added to the event?

File: src/carousel.ts [hunk 5]
```diff
@@ -22,6 +26,34 @@ function carousel(id: string, curChild=0) {
...
+            if(type === 'swiperight'){
+                if(curChild === 0){
+                    return;
+                }
+                centerCard(--curChild);
+                showControls();
+            }
+            //Check if user swipes left (image moves left)
+            if(type === 'swipeleft'){
+                if(curChild === numChildren - 1){
```
Comment (mckoss) [inline src/carousel.ts]: scrollBy(1)

**Review by mckoss — CHANGES_REQUESTED** [review]
I've added some lint checks too for missing semicolons.  And fixed the files that had errors.

File: src/carousel.ts [hunk 6]
```diff
@@ -22,6 +26,34 @@ function carousel(id: string, curChild=0) {
...
 
+
+    //Enable Swipe on mobile devices only
+    hammer.set({enable: windowSize <= 768 ? true : false});
+    //Detect swipe left or right on divParent
+    hammer.on("swipeleft swiperight", function(e: any) {
+        const {type, isFinal} = e;
+
+        //Event listener fires mutiple times. Waits until events are done.
+        if(isFinal){
```
Comment (mckoss) [inline src/carousel.ts]: I want to enforce space after "if" - I added a lint rule to pick these up.

**Comment by mckoss** [comment]
@AlexMNet I've left some review comments - and added some lint rules that are not currently passing.

You'll have to pull this branch and then continue to pick up my changes to your branch.

**Comment by AlexMNet** [comment]
@mckoss Changes have been made!

File: src/carousel.ts [hunk 7]
```diff
@@ -22,21 +26,25 @@ function carousel(id: string, curChild=0) {
...
-        }
-        centerCard(--curChild);
-        showControls();
-    });
 
-    divRight.addEventListener('click', () => {
-        if (curChild === numChildren - 1) {
-            return;
+    //Enable Swipe on mobile/tablet devices only
+    hammer.set({enable: windowSize <= 1024 ? true : false});
```
Comment (mckoss) [inline src/carousel.ts]: Why not get rid of this altogether?  And aren't some iPad dimensions over 2000 pixels?  Not to mention other devices with touch screens.  This doesn't hurt to have this library loaded on all devices, does it?

**Review by mckoss — CHANGES_REQUESTED** [review]
I made a few edits again.  We don't allow if's w/o curly body as a style convention (I added another eslint rule to catch it).

You'll have to pull again.

Also - types are dev dependencies, not runtime dependencies.  I fixed the TS error.

Last issue is removing the screen width check if you think that's ok.

File: src/carousel.ts (hunk 7 above)
Comment (AlexMNet) [reply src/carousel.ts]: It actually works quite well on desktop by clicking and dragging. Only issue is if a user tries to click and drag on the image it drags a ghost image and it doesn't swipe. I believe the ghost image issue should be an easy fix though and can create another issue for that and handle.

File: src/carousel.ts [hunk 8]
```diff
@@ -22,22 +25,24 @@ function carousel(id: string, curChild=0) {
...
-        showControls();
-    });
+    //Detect swipe left or right on divParent
+    hammer.on("swipeleft swiperight", (e: HammerInput) => {
+        const { type, isFinal } = e;
 
-    divRight.addEventListener('click', () => {
-        if (curChild === numChildren - 1) {
-            return;
+        //Event listener fires multiple times. Waits until events are done.
```
Comment (mckoss) [inline src/carousel.ts]: Space after comment //

File: src/carousel.ts [hunk 9]
```diff
@@ -53,4 +58,16 @@ function carousel(id: string, curChild=0) {
         divLeft.style.visibility = curChild === 0 ? 'hidden' : 'visible';
         divRight.style.visibility = curChild === numChildren - 1 ? 'hidden' : 'visible';
     }
+
+    function scrollBy(n: number) {
+        const nextChild = curChild + n;
+        console.log("nuChildren:", numChildren);
```
Comment (mckoss) [inline src/carousel.ts]: Don't leave console.log's in

**Review by mckoss — APPROVED** [review]
I'll merge this one - good job!

**Pull request merged**